test/Microsoft.AspNet.Routing.Tests/Template/TemplateRouteTest.cs

```diff
@@ -1352,7 +1280,126 @@ public void GetVirtualPath_InlineConstraints_CompositeConstraint_FromConstructor
             Assert.Equal("Home/Index/products", path);
         }
 
-        
+        [Fact]
+        public void GetVirtualPath_OptionalParameter_ParameterPresentInValues()
+        {
+            // Arrange            
+            var route = CreateRoute(
+                template: "{controller}/{action}/{name}.{format?}",
+                defaults: null,
+                accept: true,
+                constraints: null);
+
+            var context = CreateVirtualPathContext(
+                values: new { action = "Index", controller = "Home", name = "products", format = "xml"});
+
+            // Act
+            var path = route.GetVirtualPath(context);
+
+            // Assert
+            Assert.Equal("Home/Index/products.xml", path);
+        }
+
+        [Fact]
+        public void GetVirtualPath_OptionalParameter_ParameterNotPresentInValues()
+        {
+            // Arrange            
+            var route = CreateRoute(
+                template: "{controller}/{action}/{name}.{format?}",
+                defaults: null,
+                accept: true,
+                constraints: null);
+
+            var context = CreateVirtualPathContext(
+                values: new { action = "Index", controller = "Home", name = "products"});
+
+            // Act
+            var path = route.GetVirtualPath(context);
+
+            // Assert
+            Assert.Equal("Home/Index/products", path);
+        }
+
+        [Fact]
+        public void GetVirtualPath_OptionalParameter_ParameterPresentInValuesAndDefaults()
+        {
+            // Arrange            
+            var route = CreateRoute(
+                template: "{controller}/{action}/{name}.{format?}",
+                defaults: new { format = "json" },
+                accept: true,
+                constraints: null);
+
+            var context = CreateVirtualPathContext(
+                values: new { action = "Index", controller = "Home", name = "products" , format = "xml"});
+
+            // Act
+            var path = route.GetVirtualPath(context);
+
+            // Assert
+            Assert.Equal("Home/Index/products.xml", path);
+        }
```

Comment:
Yes, I wanted to talk to you about this test case, I had this one and wasn't sure why it was omitted from the link. Is it because it is ambient value and gets invalidated...